src/biconnectivity13/biconnect.jl

```diff
@@ -84,16 +27,72 @@ julia> biconnected_components(cycle_graph(5))
 """
 function biconnected_components end
 @traitfn function biconnected_components(g::::(!IsDirected))
-    state = Biconnections(g)
-    for u in vertices(g)
-        if state.depth[u] == 0
-            visit!(g, state, u, u)
-        end
+        n = nv(g)
+        S = Array{Tuple{Int, Int}, 1}(undef,n)
+        E = Edge{eltype(g)}
+        edge_st = Vector{E}()
+        components = Vector{Vector{E}}()
+        idx = zero(Int)      # preorder counter
+        low = zeros(Int, nv(g))
+        pre = zeros(Int, nv(g))
+        #stack pointer to emulate what happens in function call,
+        #that will make us avoid the reallocation of push and pop vector
+        stack_ptr = zero(Int) #stack pointer to emulate what happens in function  call, that will make us avoid the reallocation of push and pop vector
+
+    @inbounds for us in vertices(g)
+        pre[us] == 0 || continue # don’t go to visited nodes again
+        stack_ptr += 1
+        S[stack_ptr] = (us, 1)
+        ptr = one(Int)
+
+        while stack_ptr > 0
+            v, _ = S[stack_ptr]
+            if ptr == 1  # if ptr == 1 then we all in this node for the first time
+                idx+=1
+                low[v] = pre[v] = idx
+            end
 
-        if !isempty(state.stack)
-            push!(state.biconnected_comps, reverse(state.stack))
-            empty!(state.stack)
+            neighs = outneighbors(g, v)
+             while ptr <= length(neighs)
+                i = neighs[ptr]
+
+                if pre[i] ==  0
+                    push!(edge_st, E(min(i, v), max(i, v)))
+                    stack_ptr += 1
+                    S[stack_ptr] = (i, ptr+1)
+                    break
+                elseif (!(stack_ptr > 1 && i == S[stack_ptr-1][1])) && pre[i] < low[v]
+                    low[v] = pre[i]
+                    push!(edge_st, E(min(v, i), max(v, i)))
+                end
+
+                ptr += 1
+            end
+
+            #if ptr > length(neighs) then i finished my processing
+            # and will return to my parent, else i will go to new node
+
+            if ptr > length(neighs)
+                _, ptr = S[stack_ptr]
+                stack_ptr -= 1
+                if stack_ptr >= 1
+                    p, _ = S[stack_ptr]   #  my parent
+                   if low[v] < low[p]
```

Comment:
indentation issues. PLEASE find an editor that fixes this for you. We can't spend any more time reviewing code that isn't formatted properly.